.maint/update_authors.py

```diff
@@ -0,0 +1,302 @@
+
+    Examples
+    --------
+    >>> read_md_table('''
+    ... # Some text
+    ...
+    ... More text
+    ...
+    ... | **Header1** | **Header2** |
+    ... | --- | --- |
+    ... | val1 | val2 |
+    ... |  | val4 |
+    ...
+    ... | **Header3** | **Header4** |
+    ... | --- | --- |
+    ... | val1 | val2 |
+    ... |  | val4 |
+    ... ''')
+    [{'header1': 'val1', 'header2': 'val2'}, {'header2': 'val4'}]
+
+    """
+    prev = None
+    keys = None
+    retval = []
+    for line in md_text.splitlines():
+        if line.strip().startswith("| --- |"):
+            keys = (k.replace("*", "").strip() for k in prev.split("|"))
+            keys = [k.lower() for k in keys if k]
+            continue
+        elif not keys:
+            prev = line
+            continue
+
+        if not line or not line.strip().startswith("|"):
+            break
+
+        values = [v.strip() or None for v in line.split("|")][1:-1]
+        retval.append({k: v for k, v in zip(keys, values) if v})
+
+    return retval
+
+
+def _extract_git_contributor_matches(git_lines, sorted_authors, first_last, first_last_excl):
+    unmatched = []
+    author_matches = []
+    for ele in git_lines:
+        matches = process.extract(ele, first_last, scorer=fuzz.token_sort_ratio, limit=2)
+        # matches is a list [('First match', % Match), ('Second match', % Match)]
+        if matches[0][1] > 80:
+            val = sorted_authors[first_last.index(matches[0][0])]
+        else:
+            # skip unmatched names
+            if ele not in first_last_excl:
+                unmatched.append(ele)
+            continue
+
+        if val not in author_matches:
+            author_matches.append(val)
+
+    return author_matches, unmatched
+
+
+def _get_position_matches(author_matches):
+    position_matches = []
+    for i, item in enumerate(author_matches):
+        pos = item.pop("position", None)
+        if pos is not None:
+            position_matches.append((i, int(pos)))
+
+    return position_matches
+
+
+def sort_contributors(entries, git_lines, exclude=None, last=None):
+    """Return a list of author dictionaries, ordered by contribution."""
+    last = last or []
+    sorted_authors = sorted(entries, key=lambda i: i["name"])
+
+    first_last = [" ".join(val["name"].split(",")[::-1]).strip() for val in sorted_authors]
+    first_last_excl = [" ".join(val["name"].split(",")[::-1]).strip() for val in exclude or []]
+
+    author_matches, unmatched = _extract_git_contributor_matches(
+        git_lines, sorted_authors, first_last, first_last_excl
+    )
+
+    names = {" ".join(val["name"].split(",")[::-1]).strip() for val in author_matches}
+    for missing_name in first_last:
+        if missing_name not in names:
+            missing = sorted_authors[first_last.index(missing_name)]
+            author_matches.append(missing)
+
+    position_matches = _get_position_matches(author_matches)
+
+    for i, pos in position_matches:
```

Comment:
Did not outsource this to another method as it was not all that clear what the state of the array is before/after this (e.g. whether the `author_matches` array is re-written entirely, etc.)